kinematic_properties.py

```diff
@@ -242,24 +240,77 @@ def get_inertia_tensor(
        Masses of the particles.
      - position: unyt.unyt_array
        Positions of the particles.
+     - radius: unyt.unyt_quantity
+       Exclude particles outside this radius for the inertia tensor calculation
+     - reduced: bool
+       Whether to calculate the reduced inertia tensor
+     - max_iterations: int
+       The maximum number of iterations to repeat the inertia tensor calculation
 
     Returns the inertia tensor.
-    """
-    # 3x3 inertia tensor
-    Itensor = (
-        mass[:, None, None] / mass.sum() * position[:, None:, None] * position[:, None]
-    ).sum(axis=0)
-
-    # Symmetric, so only return lower triangle
-    Itensor = np.concatenate([np.diag(Itensor), Itensor[np.triu_indices(3, 1)]])
-
-    return Itensor
-
-
-def get_projected_inertia_tensor(
-    mass: unyt.unyt_array, position: unyt.unyt_array, axis: int
-) -> unyt.unyt_array:
-    """
+    '''
+    # Shouldn't calculate an inertia tensor for small numbers of particles
+    if position.shape[0] < 20:
+        return None
+
+    # Remove particles at centre if calculating reduced tensor
+    if reduced:
+        norm = np.linalg.norm(position, axis=1) ** 2
+        mask = np.logical_not(np.isclose(norm, 0))
+        position = position[mask]
+        mass = mass[mask]
+        norm = norm[mask]
+
+    # Set stopping criteria
+    tol = 0.0001
+    q = 1000
+
+    # Ensure we have consistent units
+    R = radius.to('kpc')
+    position = position.to('kpc')
+
+    # Start with a sphere
+    eig_val = [1, 1, 1]
+    eig_vec = np.array([
+        [1, 0, 0],
+        [0, 1, 0],
+        [0, 0, 1],
+    ])
+
+    for _ in range(max_iterations):
+        # Calculate shape
+        old_q = q
+        q = np.sqrt(eig_val[1] / eig_val[2])
+        s = np.sqrt(eig_val[0] / eig_val[2])
+        p = np.sqrt(eig_val[0] / eig_val[1])
+
+        # Break if converged
+        if abs((old_q - q) / q) < tol:
+            break
+
+        # Calculate ellipsoid, determine which particles are inside
+        axis = R * np.array([
```

Comment:
Do you remember where you got these formulaes from? I would like to double check the expressions.

Reply:
I derived them. The volume is proportional to the product of the three axis lengths. Any axis length is proportional to the other axis lengths based on the axis ratios which we have from the eigenvalues. I get these expressions by equating the volume of the initial sphere with the volume as a function of a single axis length.